pkg/events/lookuptable.go

```diff
@@ -0,0 +1,105 @@
+package events
+
+import (
+	"errors"
+	"fmt"
+)
+
+type EventLookup interface {
+	// Get retrieves the event specified the reference. If the event doesn't exist, nil is returned.
+	Get(ref Ref) Event
+	// FindLastEvent finds the last event in the event path which matches the specified matcher. If there are multiple
+	// event paths that match, an error is returned. If no events match, nil is returned.
+	FindLastEvent(matcher EventMatcher) (Event, error)
+}
+
+type eventLookupTable struct {
+	// refs contains all event references from parent to child (given that B refers to previous event A; {A -> B})
+	refs map[Event]Event
+	// entries contains all events indexed by their ref {(ref(A) -> A, ref(B) -> B}
+	entries map[string]Event
+}
+
+func newEventLookupTable() *eventLookupTable {
+	return &eventLookupTable{refs: make(map[Event]Event, 0), entries: make(map[string]Event, 0)}
+}
+
+func (r eventLookupTable) Get(ref Ref) Event {
+	return r.entries[ref.String()]
+}
+
+func (r eventLookupTable) FindLastEvent(matcher EventMatcher) (Event, error) {
```

Comment:
Is it time for sqlite yet or .... ?

Reply:
This stuff is actually quite hard to do in SQL-based databases (walking trees/sequences, requiring recursiveness) and/or require stored-procedures or complex database-specific queries, so I think this is simpler (and faster) for now.

Reply:
What about https://github.com/etcd-io/bbolt?

Reply:
I think we should store the events in sqlite (since that's the database already in place for Nuts). Then is the question whether derived data (like this event lookup table) should also be stored, or just kept in-memory (as currently implemented). Or maybe as a view in sqlite.

The events are the source data, I'd only persist derived data when required for performance or recuded complexity.